drivers/charger/emul_sbs_charger.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Emulator for SBS 1.1 compliant smart battery charger.
+ */
+
+#define DT_DRV_COMPAT sbs_sbs_charger
+
+#include "sbs_charger.h"
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(sbs_sbs_charger);
+
+/** Static configuration for the emulator */
+struct sbs_charger_emul_cfg {
+	/** I2C address of emulator */
+	uint16_t addr;
+};
+
+static int emul_sbs_charger_reg_write(const struct emul *target, int reg, int val)
+{
+	LOG_INF("write %x = %x", reg, val);
+	switch (reg) {
+	default:
+		LOG_ERR("Unknown write %x", reg);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int emul_sbs_charger_reg_read(const struct emul *target, int reg, int *val)
+{
+	switch (reg) {
+	case SBS_CHARGER_REG_SPEC_INFO:
+	case SBS_CHARGER_REG_STATUS:
+	case SBS_CHARGER_REG_ALARM_WARNING:
+		/* Arbitrary stub value. */
+		*val = 1;
+		break;
+	default:
+		LOG_ERR("Unknown register 0x%x read", reg);
+		return -EIO;
+	}
+	LOG_INF("read 0x%x = 0x%x", reg, *val);
+
+	return 0;
+}
+
+static int sbs_charger_emul_transfer_i2c(const struct emul *target, struct i2c_msg *msgs,
+					 int num_msgs, int addr)
+{
+	/* Largely copied from emul_sbs_gauge.c */
+	struct sbs_charger_emul_data *data;
+	unsigned int val;
+	int reg;
+	int rc;
+
+	data = target->data;
+
+	i2c_dump_msgs_rw("emul", msgs, num_msgs, addr, false);
+	switch (num_msgs) {
+	case 2:
+		if (msgs->flags & I2C_MSG_READ) {
+			LOG_ERR("Unexpected read");
+			return -EIO;
+		}
+		if (msgs->len != 1) {
+			LOG_ERR("Unexpected msg0 length %d", msgs->len);
+			return -EIO;
+		}
+		reg = msgs->buf[0];
+
+		/* Now process the 'read' part of the message */
+		msgs++;
+		if (msgs->flags & I2C_MSG_READ) {
+			switch (msgs->len - 1) {
+			case 1:
+				rc = emul_sbs_charger_reg_read(target, reg, &val);
+				if (rc) {
+					/* Return before writing bad value to message buffer */
+					return rc;
+				}
+
+				/* SBS uses SMBus, which sends data in little-endian format. */
+				sys_put_le16(val, msgs->buf);
```

Comment:
Should this emulator use the `include/zephyr/drivers/smbus.h` API instead, then?

Reply:
@henrikbrixandersen 

> Should this emulator use the include/zephyr/drivers/smbus.h API instead, then?

I don't think so. It seems the `smbus.h` header provides utilities like `smbus_word_data_read()` for drivers to talk on an SMBus peripheral. I don't think that's really useful for the emulator here.

Reply:
@henrikbrixandersen, so technically yes it should, but this device doesn't really make extensive use of the SMBus commands. I2C with byte manipulation suffices for the purposes of what we are trying to do here.

Additionally, this hardware is pretty old and hard to find, so I am relying on the I2C emulator to test this API and driver. I do not have the bandwidth to develop/test/upstream an SMBus emulator to then test this driver. I think @aaronemassey may be considering an SMBus emulator down the line, but without it I think this change introduces some significant risk. This PR is going on a quarter of year now and changing to the SMBus API may set us back even farther.

Reply:
Really the SMBus emulator would be another bus emulator node similar to `zephyr,i2c-emul-controller.yaml`, I think for our purposes within this PR it's not terribly useful at the moment but it's something I could create that would help out with [emulator boiler-plate for SMBus peripheral emulators.](https://github.com/zephyrproject-rtos/zephyr/issues/59211)

Reply:
@henrikbrixandersen, what are your thoughts?

Reply:
Okay, that sounds fair.